### PR TITLE
Logging: Remove dbClient global variable

### LINKED ISSUE #2070: support-logging: refactor to remove dbClient global variable
Refactor the above-identified service to remove the above-identified global variable (which is defined in the service's init.go).

This work has already been completed in the system management agent.  Use it as a reference.  See https://github.com/edgexfoundry/edgex-go/blob/master/internal/system/agent/router.go and https://github.com/edgexfoundry/edgex-go/blob/master/internal/system/agent/init.go.

You will need to update route creation to explicitly pass the individual required dependencies to the controller functions.  This will bubble down through the implementation; that is, you'll need to pass those dependencies down the call chain to where they're currently referencing the global variable.  Do not pass the dependency injection container into any controller functions.

You can also create any long-lived dependencies in the service's bootstrap and use the dependency injection container to hold them.  This was done in the system management agent; look there for an example.

Add me as the sole reviewer of your PR.

Note this is Geneva work and should not be merged into master prior to Fuji branching.  As there are a series of these issues for this service (each refactoring away a global variable), assign all of this service's related issues to you and use the branch created for this PR as the starting point for your next PR to avoid extensive conflicts. 

Depends on https://github.com/edgexfoundry/edgex-go/issues/2069

### PR DESCRIPTION
**Pull request by brandonforster**
Fixes #2070 

- Refactored support-logging by creating two new interfaces: `Criteria` and `Persistence`. These interfaces are necessary because the container needs an external interface to reference (`Persistence`) and `Persistence` depends on a parameter called a `MatchCriteria` which then itself needs an interface to not have a dependency on the main `logging` package.

- Moved all non-superflous controller logic out of `router.go`. Since this is a huge change and the primary driver of the source lines of code change, I have made one commit where I move code without changing anything, and the second commit with functional changes. This should allow the changes to be parsed.

- Removed `dbClient` `Persistence` global.

To verify, run the unit and black box tests.

Signed-off-by: Brandon Forster <brandonforster@gmail.com>

### CODE REVIEW

File: internal/support/logging/interfaces/persistence.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+type Persistence interface {
+	Add(logEntry models.LogEntry) error
+	CloseSession()
+	Remove(criteria Criteria) (int, error)
+	Find(criteria Criteria) ([]models.LogEntry, error)
+
+	// Needed for the tests. Reset the instance (closing files, sessions...)
+	// and clear the logs.
+	Reset()
```
Comment (AnthonyMBonafide) [inline internal/support/logging/interfaces/persistence.go]: Do we need to have this method in the interface, and/or is there a better way to achieve this? My concerns here are:  

1. We are now requiring implementations to provide functionality that is only useful during testing.
2. The perceived intention of this function is to purge all data to allow for a clean testing slate. An accidental call in production is more likely and unrecoverable, that is if the implementation is indeed purging data. For example, the Mongo implementation will permanently delete everything.

File: internal/support/logging/mongo.go [hunk 2]
```diff
@@ -124,29 +129,32 @@ func (ml *mongoLog) remove(criteria matchCriteria) (int, error) {
...
+	le := []models.LogEntry{}
+	if err := q.Limit(matchCriteria.Limit).All(&le); err != nil {
 		return le, err
 	}
 
 	return le, nil
 }
 
-func (ml *mongoLog) reset() {
+func (ml *mongoLog) Reset() {
```
Comment (AnthonyMBonafide) [inline internal/support/logging/mongo.go]: See my last comment regarding this now being exported and more accessible.

File: internal/support/logging/interfaces/persistence.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+const (
+	PersistenceDB   = "database"
+	PersistenceFile = "file"
+)
+
+type Persistence interface {
+	Add(logEntry models.LogEntry) error
+	CloseSession()
+	Remove(criteria Criteria) (int, error)
+	Find(criteria Criteria) ([]models.LogEntry, error)
```
Comment (AnthonyMBonafide) [inline internal/support/logging/interfaces/persistence.go]: Can we add GoDocs to these exported entities?

**Comment by michaelestrin** [comment]
> These interfaces are necessary because the container needs an external interface

FWIW the dependency injection container does not require interfaces; it does support concrete types.

**Comment by brandonforster** [comment]
> > These interfaces are necessary because the container needs an external interface
> 
> FWIW the dependency injection container does not require interfaces; it does support concrete types.

Right, but before this PR, there is an unexported interface and two implementations both in the `logging` package. There is no way to reference that interface in the container without a cyclical dependency unless the interface is moved to its own package. That's what I was trying to say; sorry I was unclear.

**Comment by michaelestrin** [comment]
There's some questions around whether or not we should be logging to a database; it's useful for development, less so for production.  Thinking here is ELK stack (or similar) would substitute for support-logging.  Ask is to hold off merging this work until I return, review, and resolve the database logging question.

**Comment by brandonforster** [comment]
@AnthonyMBonafide @michaelestrin Rebased and feedback addressed.

**Comment by michaelestrin** [comment]
Approved.  Setting aside until https://github.com/edgexfoundry/edgex-go/pull/2196 is merged.  Expect merge conficts that will require additional work.  Closing PR to ensure it is not accidentally merged.  Will use this work for follow-on PR that is rebased over merged #2196 and that addresses the anticipated merge conflicts.

**Pull request closed without merging**